lib/Conversion/TorchToLinalg/DataMovement.cpp

```diff
@@ -334,139 +403,200 @@ class ConvertAtenViewOp : public OpConversionPattern<AtenViewOp> {
+    //      e.g. [-1, -1] -> [2, 3, 4]
+    // 3. Set inputShapeVec and outputShape following the requirements by
     // tensor.expand_shape verification code:
     //    a. As long as one or more of the related dimensions in the expanded
     //    shape is dynamic the collapsed dimension is dynamic.
     //    b. If all of the related dimensions are static, the collapsed
     //    dimension must be static. In other words, if a collapsed dimension is
     //    dynamic, at least one of the related dimensions need to be dynamic.
-    int64_t collapsedDim = 0, expandedDim = 0;
-    while (collapsedDim < collapsedRank && expandedDim < expandedRank) {
-      // Not empty means the associations has been filled in and the dimension
-      // is unchanged.
-      if (!reassociation[collapsedDim].empty()) {
-        if (expandedDim != reassociation[collapsedDim][0])
-          return op.emitOpError("Unsupported: expanded dims are off from the "
-                                "expected dim got from reassociation");
-        collapsedDim++;
-        expandedDim++;
-        continue;
-      }
-
-      // Collect the dims that are collapsed until hitting the next dim that's
-      // unchanged.
-      SmallVector<int64_t> collapsedDims;
-      while (collapsedDim < collapsedRank &&
-             reassociation[collapsedDim].empty()) {
-        collapsedDims.push_back(collapsedDim);
-        collapsedDim++;
-      }
-      // the next reassociation is for a dim that's unchanged.
-      int64_t expandedDimNext = collapsedDim != collapsedRank
-                                    ? reassociation[collapsedDim][0]
-                                    : expandedRank;
-      if (collapsedDims.size() == 1) {
-        int64_t collapsedDimSize = 1;
-        int64_t collapsedDim = collapsedDims[0];
-        for (auto i : llvm::seq<int64_t>(expandedDim, expandedDimNext)) {
-          reassociation[collapsedDim].push_back(i);
-          if (collapsedDimSize == kUnknownSize)
-            continue;
-
-          int64_t expandedDimSize = expandedShape[i];
-          if (expandedDimSize == kUnknownSize) {
-            collapsedDimSize = kUnknownSize;
-            continue;
+    int64_t inputDim = 0, outputDim = 0;
+    for (auto boundary : unchangedDims) {
+      // We assume dims specified by AtenSizeInt ops are unchanged
+      int64_t nextUnchangedInput = boundary[0];
+      int64_t nextUnchangedOutput = boundary[1];
+
+      bool hasDynamic = false;
+      while (inputDim < nextUnchangedInput && outputDim < nextUnchangedOutput) {
+        inputAssociations.emplace_back();
+        outputAssociations.emplace_back();
+
+        // outputDim is next to the boundary
+        if (outputDim == nextUnchangedOutput - 1) {
+          if (hasDynamic && inputDim != nextUnchangedInput - 1) {
+            return rewriter.notifyMatchFailure(
+                op,
+                "found ambiguous association of dynamic sizes");
```

Comment:
nit: can you have a more descriptive error message? Namely, I think it would be useful to mention that this error happened because `inputDim` did not have the value that was expected. This allows developers to easily find the line where a match failure is generated (currently this error message is identical to one used below)